pytest.xfail reported as hard failure

I was testing out some of the error reporting and I noticed that apparently ``pytest.xfail()`` is reported as a hard failure when called within a subtest:

With this in `test.py`:

```python
 import pytest

def test_xfail(subtests):
    with subtests.test():
        pytest.xfail()
```

I get:

```
$ pytest -v test.py
...
test.py::test_a_bunch_of_stuff FAILED    [100%]
test.py::test_a_bunch_of_stuff PASSED    [100%]

============ FAILURES =========
____ test_xfail (<subtest>) ___

subtests = SubTests(ihook=<pluggy.hooks._HookRelay object at 0x7f1539aafee0>,
                    suspend_capture_ctx=<bound method CaptureManager.gl...ed'
                    _in_suspended=False> _capture_fixture=None>>,
                    request=<SubRequest 'subtests' for
                    <Function test_a_bunch_of_stuff>>)

    def test_a_bunch_of_stuff(subtests):
        with subtests.test():
>           pytest.xfail()
E           _pytest.outcomes.XFailed: <XFailed instance>

test.py:27: XFailed
========= short test summary info =========
FAILED test.py::test_a_bunch_of_stuff - _pytest.outcomes.XFailed: <XFailed instance>
====== 1 failed, 1 passed in 0.03s ========
```

Without the subtest, the overall test is properly reported as XFAIL.

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -1,6 +1,13 @@
 CHANGELOG
 =========
 
+0.5.0 (2021-05-29)
+------------------
+
+* Add support for ``pytest.mark.xfail`` (`#40`_).
+
+.. _#40: https://github.com/pytest-dev/pytest-subtests/pull/40
+
 0.4.0 (2020-12-13)
 ------------------
 

--- a/pytest_subtests.py
+++ b/pytest_subtests.py
@@ -71,6 +71,10 @@ class SubTestReport(TestReport):
         )
         return report
 
+    @classmethod
+    def from_test_report(cls, test_report):
+        return super()._from_json(test_report._to_json())
+
 
 def _addSubTest(self, test_case, test, exc_info):
     if exc_info is not None:
@@ -78,7 +82,8 @@ def _addSubTest(self, test_case, test, exc_info):
         call_info = make_call_info(
             ExceptionInfo(exc_info), start=0, stop=0, duration=0, when="call"
         )
-        sub_report = SubTestReport.from_item_and_call(item=self, call=call_info)
+        report = self.ihook.pytest_runtest_makereport(item=self, call=call_info)
+        sub_report = SubTestReport.from_test_report(report)
         sub_report.context = SubTestContext(msg, dict(test.params))
         self.ihook.pytest_runtest_logreport(report=sub_report)
         if check_interactive_exception(call_info, sub_report):
@@ -171,7 +176,8 @@ class SubTests(object):
         call_info = make_call_info(
             exc_info, start=start, stop=stop, duration=duration, when="call"
         )
-        sub_report = SubTestReport.from_item_and_call(item=self.item, call=call_info)
+        report = self.ihook.pytest_runtest_makereport(item=self.item, call=call_info)
+        sub_report = SubTestReport.from_test_report(report)
         sub_report.context = SubTestContext(msg, kwargs.copy())
 
         captured.update_report(sub_report)

--- a/pytest_subtests.py
+++ b/pytest_subtests.py
@@ -72,7 +72,7 @@ class SubTestReport(TestReport):
         return report
 
     @classmethod
-    def from_test_report(cls, test_report):
+    def _from_test_report(cls, test_report):
         return super()._from_json(test_report._to_json())
 
 
@@ -83,7 +83,7 @@ def _addSubTest(self, test_case, test, exc_info):
             ExceptionInfo(exc_info), start=0, stop=0, duration=0, when="call"
         )
         report = self.ihook.pytest_runtest_makereport(item=self, call=call_info)
-        sub_report = SubTestReport.from_test_report(report)
+        sub_report = SubTestReport._from_test_report(report)
         sub_report.context = SubTestContext(msg, dict(test.params))
         self.ihook.pytest_runtest_logreport(report=sub_report)
         if check_interactive_exception(call_info, sub_report):
@@ -177,7 +177,7 @@ class SubTests(object):
             exc_info, start=start, stop=stop, duration=duration, when="call"
         )
         report = self.ihook.pytest_runtest_makereport(item=self.item, call=call_info)
-        sub_report = SubTestReport.from_test_report(report)
+        sub_report = SubTestReport._from_test_report(report)
         sub_report.context = SubTestContext(msg, kwargs.copy())
 
         captured.update_report(sub_report)

--- a/tests/test_subtests.py
+++ b/tests/test_subtests.py
@@ -103,6 +103,27 @@ class TestFixture:
         expected_lines += ["* 1 passed, 3 skipped in *"]
         result.stdout.fnmatch_lines(expected_lines)
 
+    def test_xfail(self, testdir, mode):
+        testdir.makepyfile(
+            """
+            import pytest
+            def test_foo(subtests):
+                for i in range(5):
+                    with subtests.test(msg="custom", i=i):
+                        if i % 2 == 0:
+                            pytest.xfail('even number')
+        """
+        )
+        if mode == "normal":
+            result = testdir.runpytest()
+            expected_lines = ["collected 1 item"]
+        else:
+            pytest.importorskip("xdist")
+            result = testdir.runpytest("-n1")
+            expected_lines = ["gw0 [1]"]
+        expected_lines += ["* 1 passed, 3 xfailed in *"]
+        result.stdout.fnmatch_lines(expected_lines)
+
 
 class TestSubTest:
     """
@@ -232,6 +253,35 @@ class TestSubTest:
             result = testdir.runpytest(p)
             result.stdout.fnmatch_lines(
                 ["collected 1 item", "* 3 skipped, 1 passed in *"]
+            )
+
+    @pytest.mark.parametrize("runner", ["unittest", "pytest-normal", "pytest-xdist"])
+    @pytest.mark.xfail(reason="Not producing the expected results (#5)")
+    def test_xfail(self, testdir, runner):
+        p = testdir.makepyfile(
+            """
+            import pytest
+            from unittest import expectedFailure, TestCase, main
+
+            class T(TestCase):
+                @expectedFailure
+                def test_foo(self):
+                    for i in range(5):
+                        with self.subTest(msg="custom", i=i):
+                            if i % 2 == 0:
+                                raise pytest.xfail('even number')
+
+            if __name__ == '__main__':
+                main()
+        """
+        )
+        if runner == "unittest":
+            result = testdir.runpython(p)
+            result.stderr.fnmatch_lines(["Ran 1 test in *", "OK (expected failures=3)"])
+        else:
+            result = testdir.runpytest(p)
+            result.stdout.fnmatch_lines(
+                ["collected 1 item", "* 3 xfailed, 1 passed in *"]
             )
 
 

